File: src/guess.nim
#
# guess.nim - guessing game in nim
#
# This is written to demonstrate this language versus the same program
# written in other languages.
#
# COMPILE:
#       nim compile guess.nim
#
# 31-Aug-2018   Scott Emmons   Created.
#

import random, strutils

randomize()

let
  answer = rand(99)+1  # For nim 0.17 or older, use random(100)+1 instead
const
  scorefile = "highscores_nim"
var
  guess = -1
  num = 0

echo("guess.nim - Guess a number between 1 and 100\n")

# Play game
while guess != answer:
  num+=1
  stdout.write("Enter guess ", num, ": ")
  guess = parseInt(readLine(stdin))
  if guess < answer:
    echo("Higher...")
  elif guess > answer:
    echo("Lower...")

echo("Correct! That took ", num, " guesses.\n")

# Save high score
stdout.write("Please enter your name: ")
let name = readLine(stdin)

try:
  let outfile = open(scorefile, fmAppend)
  write(outfile, name & " " & $(num) & "\n")
  close(outfile)
except IOError:
  echo("ERROR: Can't write to ", scorefile)

# Print high scores
echo("\nPrevious high scores:")
try:
  let infile = open(scorefile)
  echo readAll(infile)
  close(infile)
except IOError:
  echo("ERROR: Can't read from ", scorefile)

File: src/guess.go
/* guess.py - guessing game in golang
*
* This is written to demonstrate this language versus the same program
* written in other languages.
*
* 22-Oct-2022   Daniel Eberhard   Created.
 */

package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"time"
)

func main() {
	scorefile := "highscores_go"
	guess := -1
	num := 0

	fmt.Println("guess.go - Guess a number between 1 and 100\n")

	// Generate random number
	rand.Seed(time.Now().UnixNano())
	answer := rand.Intn(99) + 1

	// Play game
	for guess != answer {
		num = num + 1
		fmt.Printf("Enter guess %d: ", num)
		_, _ = fmt.Scanf("%d\n", &guess)
		if guess < answer {
			fmt.Println("Higher...")
		} else if guess > answer {
			fmt.Println("Lower...")
		}
	}

	fmt.Println("Correct! That took", num, "guesses.\n")

	// Save high score
	fmt.Print("Please enter your name: ")
	var name string
	_, err := fmt.Scanf("%s\n", &name)
	outfile, err := os.OpenFile(scorefile, os.O_APPEND, 0666)
	if err != nil {
		outfile, err = os.OpenFile(scorefile, os.O_CREATE, 0666)
		if err != nil {
			fmt.Println("ERROR: Can't write to", scorefile)
		}
	}
	defer outfile.Close()
	if _, err = outfile.WriteString(fmt.Sprintf("%s %d\n", name, num)); err != nil {
		fmt.Println("ERROR: Can't write to", scorefile)
	}

	// Print high scores
	fmt.Println("\nPrevious high scores,")
	infile, err := os.Open(scorefile)
	if err != nil {
		fmt.Println("ERROR: Can't read from", scorefile)
	}
	defer infile.Close()
	lines, err := ioutil.ReadAll(infile)
	fmt.Printf("%s", lines)
}
